Add Permission Group management CLI

ecton opened the pull request:
* Added a new dependency to print command line tables. Seems to be the most popular option on crates.io, and works nicely.
* Refactored how the Account command line interface works to keep account-specific commands and parameters all in one file.
* Used the same pattern to add a PermissionGroup command line interface

daxpedda reviewed (changes requested):
Some integration tests would be nice, since we talked about that.

Here are maybe some helpful pointers:
https://rust-cli.github.io/book/tutorial/testing.html

ecton commented:
This is sadly a lot more complicated than simple integration tests, due to needing to ensure that each test run doesn't leave data behind in the database. Essentially each call to `database::pool()` in src/cli/permission_groups.rs's handle_command need to be done inside of a transaction. You might suggest changing handle_command to take a transaction, but now we're breaking apart the CLI surface from the testing surface again.

So, what I will agree with you is that the new database methods I added for this merge should be unit tested. Unfortunately try as I might, I can't come up with a way to add testing to these "simple" database commands without introducing a new CLI surface that would still be untestable. The inner contents of handle_command are basically a initialize the environment, a match statement, a query, and a print. If we move "initialize the environment" out so that we can pass a transaction in instead, that layer above is still untestable. It also means that commands could never be written that executed multiple transactions, a limitation I'm not willing to accept if we were to try to make it generic across the entire CLI surface.

The alternate is to not worry about transactions and instead have cleanup code that runs at the start of each transaction, hunting for names we use in our test cases. I really don't want to do that either.

Do you have any ideas on other approaches to take for this, @daxpedda? My personal preference would be to have a coding standard that CLI commands are very thin wrappers around unit-tested code. Anything more complicated than what we see in our current CLI should be rejected and refactored so that the CLI surface is basically a glorified switch statement.

daxpedda commented:
> My personal preference would be to have a coding standard that CLI commands are very thin wrappers around unit-tested code.

I absolutely I agree with this, whatever the CLI does, should also be doable by a library that can be consumed independent of the CLI.

Nevertheless I think we should really end-to-end test any security sensitive things, like this CLI for example.
I think the simplest way to test this is just to set-up a local database and connect to that, that way we won't have to change any code.

I understand that this is a hassle for testers, but that's fine, tests can run in CI for something like that. In any case, there will always be tests that will require a database setup, otherwise we will have no end-to-end tests at all.

Forgive me, I didn't understand everything you said, so I might be suggesting something that you already adressed.

ecton commented:
I've pushed tests for the database methods I added that weren't tested before. I'm working on refactoring the database tests to allow for integration tests in a pretty easy way, but I'm expecting these changes to "muddy" the clarity of the pull request. I'd rather you sign off on these changes separately since the refactor of the way database tests are done is an unrelated task.

Also, no need to apologize ever for any language misunderstandings. I'm sure I could have written things more clearly. We're on the same page -- you pushed me to think of a way to make testing not as ugly and more feasible.

ecton commented:
I missed adding this repo to the CodeCov app, as I didn't give it access to all repos by default. [Here's a link to the branch's coverage chart which shows the small increase](https://codecov.io/gh/khonsulabs/cosmicverge/branch/permissions-cli). In the future reports should be shown inline like they are on Kludgine.

daxpedda commented on database/src/schema/permission_group.rs:
```diff
@@ -1,27 +1,37 @@
-        name: String,
+    pub async fn create<
+        'e,
+        E: sqlx::Executor<'e, Database = sqlx::Postgres>,
```
On a side-note, we should definitely improve this pattern, instead of passing the executor all the time, we should rly make a `trait` on executor with all these functions, or better yet, wrap the executor in a `struct` and put all functionality into methods.

daxpedda commented on database/src/schema/permission_group.rs:
```diff
@@ -110,3 +141,24 @@ impl Statement {
     }
 }
+
+#[cfg(test)]
+mod tests {
```
On a general note, I don't think we should do a `mod` for tests, this also pollutes the test-output.
What do you think?

ecton replied:
I thought I was following [the recommended setup](https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html)?

The benefits of the mod, from my viewpoint:

* If you only use an import in `#[cfg(test)]`, you must have a separate using statement. If this is outside of the method, you end up having some pretty gnarly headers in your file due to having to try to understand what is included and what isn't. If you move the usings inside of the test, then we potentially repeat usings across multiple tests.
* You only have to write `#[cfg(test)]` once

I also can say I've never even thought about it when looking at the test output.

daxpedda replied:
- Agreed
- `#[test]` implies `#[cfg(test)]`
- When you have a bunch of tests failing, it helps, generally not really worth the tradeoff if there is any

I guess it's situational, personally I skip it unless I have extra `use` statements that I only need for `cfg(test)`.

ecton replied:
As per Discord I'm going to revisit these thoughts while doing changes to #17.

daxpedda commented on server/src/cli/accounts.rs:
```diff
@@ -1,43 +1,74 @@
     }
 }
+
+fn print_account(account: Account) -> std::io::Result<()> {
```
There is a derive macro, so you don't have to implement it by hand: https://docs.rs/cli-table/0.4.5/cli_table/derive.Table.html, also has all sorts of options.

This should be a method at minimum and `Display` at best, for the latter I wouldn't know what to do with the `Result` though.

ecton replied:
Are we not concerned about the impact of proc macros on compilation performance? ... but... in a point to you, I didn't set `default-featulres = false` so the feature was enabled already. I was originally avoiding the proc macro for that reason, even though I forgot to disable it.

daxpedda replied:
Personally I would never do the tradeoff between compile time and code quality, if a proc macro increases code quality, I won't care about compile time.

daxpedda commented on server/src/cli/permission_groups.rs:
```diff
@@ -0,0 +1,106 @@
+    Ok(())
+}
+
+fn print_group(group: PermissionGroup) -> std::io::Result<()> {
```
Same as above.

daxpedda commented on server/src/cli/permission_groups.rs:
```diff
@@ -0,0 +1,106 @@
+    RemoveService { service: Service },
+}
+
+pub async fn handle_command(group_command: Command) -> anyhow::Result<()> {
```
On that note, I belive this sould be a method too, what do you think?
Personally I prefer functions to not lie around if they can only be used together with a specific `struct` or `trait`, then they should be part of them, even as associated methods.

ecton replied:
I just think of how function is being called and the pattern I'm trying to achieve. Can this be done with traits? Yes, but it's just extra lines of code. In the end, the consumer will be doing:

````
PermissionGroups::handle_command(command).await
or
permission_groups::handle_command(command).await
```

which to me don't really "feel" different. Additionally, I don't love the idea of another type named PermissionGroups clogging up our Search by Symbol in our Editors, but I'm ok with a namespaced permission_groups mod.

daxpedda replied:
I meant (incl. questionable bikeshedding):
```rust
command.handle()
```
But I can see why you would like that command to be in `PermissionGroups`, I agree in that case though, that wouldn't make any sense.

daxpedda commented on database/src/schema/permission_group.rs:
```diff
@@ -106,14 +107,24 @@ impl PermissionGroup {
+/// The signature of &Option<String> is required due to how the `cli_table::Table`
+/// macro works
+fn display_permission(value: &Option<String>) -> String {
+    value.clone().unwrap_or_else(|| String::from("*"))
```
```suggestion
    value.as_deref().unwrap_or("*").to_owned()
```
